services/venice-router/src/main/java/com/linkedin/venice/router/api/VeniceVersionFinder.java

```
@@ -98,88 +99,102 @@ public int getVersion(String storeName, BasicFullHttpRequest request) throws Ven
+        }
       }
+      stats.recordStale(metadataCurrentVersionNumber, existingVersionNumber);
+      return existingVersionNumber;
+    }
 
-      /**
-       * When the router has only one available version, despite offline partitions, or dictionary not yet downloaded,
-       * it will return it as the available version.
-       * If the partitions are still unavailable or the dictionary is not downloaded by the time the records needs to
-       * be decompressed, then the router will return an error response.
-       */
-      String message = errorMessage + " Switching to serve new active version.";
-      if (!EXCEPTION_FILTER.isRedundantException(message)) {
-        LOGGER.warn(message);
+    // existing version not ready to serve -> no version ready to serve
+    if (LOGGER.isDebugEnabled()) {
+      String errorMessage = "Unable to serve new version: " + newVersionKafkaTopic + ".";
+      errorMessage += newVersionPartitionResourcesReady ? "" : " Partition resources not ready for new version.";
+      errorMessage += newVersionDecompressorReady ? "" : " Decompressor not ready for new version.";
+      errorMessage += "Unable to serve existing version: " + existingVersionNumber + " No version ready to serve.";
+      errorMessage += existingVersionStatusOnline ? "" : " Existing version has status: " + existingVersionStatus;
+      errorMessage += existingVersionDecompressorReady
+          ? ""
+          : " Decompressor not ready for existing version. (Has dictionary downloaded?)";
+      if (!EXCEPTION_FILTER.isRedundantException(errorMessage)) {
+        LOGGER.debug(errorMessage);
       }
-      lastCurrentVersionMap.put(storeName, newCurrentVersion);
-      stats.recordNotStale();
-      return newCurrentVersion;
     }
+    stats.recordStale(metadataCurrentVersionNumber, Store.NON_EXISTING_VERSION);
+    lastCurrentVersionMap.put(storeName, Store.NON_EXISTING_VERSION);
+    return Store.NON_EXISTING_VERSION;
```

Why are we diverging from previous behavior of 

       * When the router has only one available version, despite offline partitions, or dictionary not yet downloaded,
       * it will return it as the available version.
       * If the partitions are still unavailable or the dictionary is not downloaded by the time the records needs to
       * be decompressed, then the router will return an error response.

---

Thank you for your question. The rationale is, if the partitions are offline, the version is not ready to serve, the router should report that version as not ready to serve and fail fast rather than kick the can down the road.

---

Discussed offline. Conclusion: if both new and existing version are not ready, serve the existing version + log store as stale. Reason: serving something is better than nothing. eaf8a7c53290abde6715757f4f63377f5fbf15c8